File: emotionstestinglibrary/src/main/java/com/university/itis/emotionstestinglibrary/VideoRecorder.java
package com.university.itis.emotionstestinglibrary;

import android.app.Activity;
import android.content.Context;
import android.graphics.Matrix;
import android.graphics.RectF;
import android.graphics.SurfaceTexture;
import android.hardware.camera2.CameraAccessException;
import android.hardware.camera2.CameraCaptureSession;
import android.hardware.camera2.CameraCharacteristics;
import android.hardware.camera2.CameraDevice;
import android.hardware.camera2.CameraManager;
import android.hardware.camera2.CameraMetadata;
import android.hardware.camera2.CaptureRequest;
import android.hardware.camera2.params.StreamConfigurationMap;
import android.media.MediaRecorder;
import android.support.annotation.NonNull;
import android.util.Size;
import android.util.SparseIntArray;
import android.view.Surface;
import android.view.TextureView;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.Semaphore;
import java.util.concurrent.TimeUnit;

/**
 * Created by Daria Popova on 05.04.18.
 */
public class VideoRecorder {

    private static final int SENSOR_ORIENTATION_DEFAULT_DEGREES = 90;
    private static final int SENSOR_ORIENTATION_INVERSE_DEGREES = 270;
    private static final SparseIntArray DEFAULT_ORIENTATIONS = new SparseIntArray();
    private static final SparseIntArray INVERSE_ORIENTATIONS = new SparseIntArray();

    static {
        DEFAULT_ORIENTATIONS.append(Surface.ROTATION_0, 90);
        DEFAULT_ORIENTATIONS.append(Surface.ROTATION_90, 0);
        DEFAULT_ORIENTATIONS.append(Surface.ROTATION_180, 270);
        DEFAULT_ORIENTATIONS.append(Surface.ROTATION_270, 180);
    }

    static {
        INVERSE_ORIENTATIONS.append(Surface.ROTATION_0, 270);
        INVERSE_ORIENTATIONS.append(Surface.ROTATION_90, 180);
        INVERSE_ORIENTATIONS.append(Surface.ROTATION_180, 90);
        INVERSE_ORIENTATIONS.append(Surface.ROTATION_270, 0);
    }

    private Activity activity;
    private boolean isInit;
    private String frontCameraId;
    private boolean isRecordingVideo;

    private TextureView textureView;
    private CameraDevice cameraDevice;
    private CameraCaptureSession previewSession;
    private Size videoSize;
    private MediaRecorder mediaRecorder;
    private CameraManager manager;

    //    private HandlerThread mBackgroundThread;
//    private Handler mBackgroundHandler;

    private Semaphore cameraOpenCloseLock = new Semaphore(1);
    private Integer sensorOrientation;
    private String nextVideoAbsolutePath;
    private CaptureRequest.Builder previewBuilder;

    private CameraDevice.StateCallback stateCallback = new CameraDevice.StateCallback() {

        @Override
        public void onOpened(@NonNull CameraDevice cameraDevice) {
            VideoRecorder.this.cameraDevice = cameraDevice;
            startPreview();
            cameraOpenCloseLock.release();
            if (null != textureView) {
                configureTransform(textureView.getWidth(), textureView.getHeight());
            }
        }

        @Override
        public void onDisconnected(@NonNull CameraDevice cameraDevice) {
            cameraOpenCloseLock.release();
            cameraDevice.close();
            VideoRecorder.this.cameraDevice = null;
        }

        @Override
        public void onError(@NonNull CameraDevice cameraDevice, int error) {
            cameraOpenCloseLock.release();
            cameraDevice.close();
            VideoRecorder.this.cameraDevice = null;
        }

    };
    private TextureView.SurfaceTextureListener surfaceTextureListener
            = new TextureView.SurfaceTextureListener() {

        @Override
        public void onSurfaceTextureAvailable(SurfaceTexture surfaceTexture,
                                              int width, int height) {
            openCamera(width, height);
        }

        @Override
        public void onSurfaceTextureSizeChanged(SurfaceTexture surfaceTexture,
                                                int width, int height) {
        }

        @Override
        public boolean onSurfaceTextureDestroyed(SurfaceTexture surfaceTexture) {
            return true;
        }

        @Override
        public void onSurfaceTextureUpdated(SurfaceTexture surfaceTexture) {
            if (!isRecordingVideo) {
                startRecordingVideo();
            }
        }

    };

    public VideoRecorder(Activity activity) {
        this.activity = activity;
        this.manager = (CameraManager) activity.getSystemService(Context.CAMERA_SERVICE);
        textureView = VideoUtils.createTextureView(activity);
        frontCameraId = VideoUtils.getFrontCameraId(manager);
    }


    public void startRecording() {
        PermissionUtils.requestPermissions(activity);
        if (!PermissionUtils.hasPermissions(activity)) return;
//        startBackgroundThread();
        if (textureView.isAvailable()) {
            openCamera(textureView.getWidth(), textureView.getHeight());
        } else {
            textureView.setSurfaceTextureListener(surfaceTextureListener);
        }
        isInit = true;
    }

    public void stopRecording() {
        if (!isInit) return;
        if (isRecordingVideo) {
            stopRecordingVideo();
        }
        closeCamera();
//        stopBackgroundThread();
    }


    private void closePreviewSession() {
        if (previewSession != null) {
            previewSession.close();
            previewSession = null;
        }
    }

    private void stopRecordingVideo() {
        isRecordingVideo = false;
        mediaRecorder.stop();
        mediaRecorder.reset();
        nextVideoAbsolutePath = null;
    }


//    private void startBackgroundThread() {
//        mBackgroundThread = new HandlerThread("CameraBackground");
//        mBackgroundThread.start();
//        mBackgroundHandler = new Handler(mBackgroundThread.getLooper());
//    }
//
//    private void stopBackgroundThread() {
//        mBackgroundThread.quitSafely();
//        try {
//            mBackgroundThread.join();
//            mBackgroundThread = null;
//            mBackgroundHandler = null;
//        } catch (InterruptedException e) {
//            e.printStackTrace();
//        }
//    }

    @SuppressWarnings("MissingPermission")
    private void openCamera(int width, int height) {
        if (null == activity || activity.isFinishing()) {
            return;
        }

        try {
            if (!cameraOpenCloseLock.tryAcquire(2500, TimeUnit.MILLISECONDS)) {
                throw new RuntimeException("Time out waiting to lock camera opening.");
            }

            CameraCharacteristics characteristics = manager.getCameraCharacteristics(frontCameraId);
            StreamConfigurationMap map = characteristics
                    .get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP);
            sensorOrientation = characteristics.get(CameraCharacteristics.SENSOR_ORIENTATION);
            if (map == null) {
                throw new RuntimeException("Cannot get available preview/video sizes");
            }
            videoSize = VideoUtils.chooseVideoSize(map.getOutputSizes(MediaRecorder.class));

            int orientation = activity.getResources().getConfiguration().orientation;
//            configureTransform(width, height);
            mediaRecorder = new MediaRecorder();
            manager.openCamera(frontCameraId, stateCallback, null);
        } catch (InterruptedException e) {
            e.printStackTrace();
        } catch (CameraAccessException e) {
            e.printStackTrace();
        }

    }

    private void closeCamera() {
        try {
            cameraOpenCloseLock.acquire();
            if (null != cameraDevice) {
                cameraDevice.close();
                cameraDevice = null;
            }
            if (null != mediaRecorder) {
                mediaRecorder.release();
                mediaRecorder = null;
            }
        } catch (InterruptedException e) {
            throw new RuntimeException("Interrupted while trying to lock camera closing.");
        } finally {
            cameraOpenCloseLock.release();
        }
    }

    private void startPreview() {
        if (null == cameraDevice || !textureView.isAvailable()) {
            return;
        }
        try {
            closePreviewSession();
            SurfaceTexture texture = textureView.getSurfaceTexture();
            assert texture != null;
            previewBuilder = cameraDevice.createCaptureRequest(CameraDevice.TEMPLATE_PREVIEW);

            Surface previewSurface = new Surface(texture);
            previewBuilder.addTarget(previewSurface);

            cameraDevice.createCaptureSession(Collections.singletonList(previewSurface),
                    new CameraCaptureSession.StateCallback() {

                        @Override
                        public void onConfigured(@NonNull CameraCaptureSession session) {
                            previewSession = session;
                            updatePreview();
                        }

                        @Override
                        public void onConfigureFailed(@NonNull CameraCaptureSession session) {
                        }
                    }, null);
        } catch (CameraAccessException e) {
            e.printStackTrace();
        }
    }

    private void updatePreview() {
        if (null == cameraDevice) {
            return;
        }
        try {
            setUpCaptureRequestBuilder(previewBuilder);
//            HandlerThread thread = new HandlerThread("CameraPreview");
//            thread.start();
            previewSession.setRepeatingRequest(previewBuilder.build(), null, null);
        } catch (CameraAccessException e) {
            e.printStackTrace();
        }
    }

    private void setUpCaptureRequestBuilder(CaptureRequest.Builder builder) {
        builder.set(CaptureRequest.CONTROL_MODE, CameraMetadata.CONTROL_MODE_AUTO);
    }

    private void configureTransform(int viewWidth, int viewHeight) {
        if (null == textureView || null == activity) {
            return;
        }
        int rotation = activity.getWindowManager().getDefaultDisplay().getRotation();
        Matrix matrix = new Matrix();
        RectF viewRect = new RectF(0, 0, viewWidth, viewHeight);
        float centerX = viewRect.centerX();
        float centerY = viewRect.centerY();

        textureView.setTransform(matrix);
    }

    private void setUpMediaRecorder() throws IOException {
        if (null == activity) {
            return;
        }
        mediaRecorder.setAudioSource(MediaRecorder.AudioSource.MIC);

        mediaRecorder.setVideoSource(MediaRecorder.VideoSource.SURFACE);
        mediaRecorder.setOutputFormat(MediaRecorder.OutputFormat.MPEG_4);
        if (nextVideoAbsolutePath == null || nextVideoAbsolutePath.isEmpty()) {
            nextVideoAbsolutePath = StorageUtils.getVideoFilePath(activity);
        }
        mediaRecorder.setOutputFile(nextVideoAbsolutePath);
        mediaRecorder.setVideoEncodingBitRate(10000000);
        mediaRecorder.setVideoFrameRate(30);
        mediaRecorder.setVideoSize(videoSize.getWidth(), videoSize.getHeight());
        mediaRecorder.setVideoEncoder(MediaRecorder.VideoEncoder.H264);
        mediaRecorder.setAudioEncoder(MediaRecorder.AudioEncoder.AAC);
        int rotation = activity.getWindowManager().getDefaultDisplay().getRotation();
        switch (sensorOrientation) {
            case SENSOR_ORIENTATION_DEFAULT_DEGREES:
                mediaRecorder.setOrientationHint(DEFAULT_ORIENTATIONS.get(rotation));
                break;
            case SENSOR_ORIENTATION_INVERSE_DEGREES:
                mediaRecorder.setOrientationHint(INVERSE_ORIENTATIONS.get(rotation));
                break;
        }
        mediaRecorder.prepare();
    }

    private void startRecordingVideo() {
        if (null == cameraDevice || !textureView.isAvailable()) {
            return;
        }
        try {
            closePreviewSession();
            setUpMediaRecorder();
            SurfaceTexture texture = textureView.getSurfaceTexture();
            assert texture != null;
            previewBuilder = cameraDevice.createCaptureRequest(CameraDevice.TEMPLATE_RECORD);
            List<Surface> surfaces = new ArrayList<>();
            Surface previewSurface = new Surface(texture);
            surfaces.add(previewSurface);
            previewBuilder.addTarget(previewSurface);

            Surface recorderSurface = mediaRecorder.getSurface();
            surfaces.add(recorderSurface);
            previewBuilder.addTarget(recorderSurface);

            cameraDevice.createCaptureSession(surfaces, new CameraCaptureSession.StateCallback() {

                @Override
                public void onConfigured(@NonNull CameraCaptureSession cameraCaptureSession) {
                    previewSession = cameraCaptureSession;
                    updatePreview();
                    activity.runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            isRecordingVideo = true;
                            mediaRecorder.start();
                        }
                    });
                }

                @Override
                public void onConfigureFailed(@NonNull CameraCaptureSession cameraCaptureSession) {
                }
            }, null);
        } catch (CameraAccessException | IOException e) {
            e.printStackTrace();
        }

    }

}

File: emotionstestinglibrary/src/main/java/com/university/itis/emotionstestinglibrary/VideoRecorder.kt
package com.university.itis.emotionstestinglibrary

import android.annotation.SuppressLint
import android.app.Activity
import android.graphics.SurfaceTexture
import android.hardware.camera2.*
import android.media.MediaRecorder
import android.os.Handler
import android.os.HandlerThread
import android.util.Size
import android.view.Surface
import android.view.TextureView
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import com.university.itis.emotionstestinglibrary.ETL.manager
import com.university.itis.emotionstestinglibrary.ETL.sensorOrientation
import com.university.itis.emotionstestinglibrary.ETL.videoSize
import com.university.itis.emotionstestinglibrary.Sensors.DEFAULT_ORIENTATIONS
import com.university.itis.emotionstestinglibrary.Sensors.INVERSE_ORIENTATIONS
import com.university.itis.emotionstestinglibrary.Sensors.SENSOR_ORIENTATION_DEFAULT_DEGREES
import com.university.itis.emotionstestinglibrary.Sensors.SENSOR_ORIENTATION_INVERSE_DEGREES
import java.util.*
import java.util.concurrent.Semaphore


class VideoRecorder(private val activity: Activity) {
    
    
    private var isInit: Boolean = false
    private var isRecordingVideo: Boolean = false
    private val textureView: TextureView
    private var cameraDevice: CameraDevice? = null
    private var previewSession: CameraCaptureSession? = null
    private var backgroundThread: HandlerThread? = null
    private var backgroundHandler: Handler? = null
    private val cameraOpenCloseLock = Semaphore(1)
    private var mediaRecorder: MediaRecorder? = null
    private var nextVideoAbsolutePath: String? = null
    private var previewBuilder: CaptureRequest.Builder? = null
    private val stateCallback = object : CameraDevice.StateCallback() {
        
        override fun onOpened(cameraDevice: CameraDevice) {
            this@VideoRecorder.cameraDevice = cameraDevice
            startPreview()
            cameraOpenCloseLock.release()
        }
        
        override fun onDisconnected(cameraDevice: CameraDevice) {
            cameraOpenCloseLock.release()
            cameraDevice.close()
            this@VideoRecorder.cameraDevice = null
        }
        
        override fun onError(cameraDevice: CameraDevice, error: Int) {
            cameraOpenCloseLock.release()
            cameraDevice.close()
            this@VideoRecorder.cameraDevice = null
        }
        
    }
    private val surfaceTextureListener = object : TextureView.SurfaceTextureListener {
        
        override fun onSurfaceTextureAvailable(surfaceTexture: SurfaceTexture,
                                               width: Int, height: Int) {
            openCamera(width, height)
//
        }
        
        override fun onSurfaceTextureSizeChanged(surfaceTexture: SurfaceTexture,
                                                 width: Int, height: Int) {
        }
        
        override fun onSurfaceTextureDestroyed(surfaceTexture: SurfaceTexture): Boolean {
            surfaceTexture.release()
            return true
        }
        
        override fun onSurfaceTextureUpdated(surfaceTexture: SurfaceTexture) {
            if (!isRecordingVideo) {
                startRecordingVideo()
            }
        }
        
    }
    
    init {
        textureView = createTextureView(activity)
    }
    
    fun startRecording() {
        requestAppPermissions(activity)
        if (!hasAppPermissions(activity)) return
        startBackgroundThread()
        if (textureView.isAvailable) {
            openCamera(textureView.width, textureView.height)
        } else {
            textureView.surfaceTextureListener = surfaceTextureListener
            startRecordingVideo()
        }
        isInit = true
    }
    
    fun stopRecording() {
        if (!isInit) return
        if (isRecordingVideo) {
            try {
                stopRecordingVideo()
            } catch (e: RuntimeException) {
                deleteVideoFile(nextVideoAbsolutePath)
            }
        }
        resetMediaRecorder()
        closeCamera()
        stopBackgroundThread()
    }
    
    private fun closeCamera() {
        cameraOpenCloseLock.acquire()
        cameraDevice?.close()
        mediaRecorder?.release()
        cameraDevice = null
        mediaRecorder = null
    }
    
    private fun closePreviewSession() {
        try {
            previewSession?.close()
            previewSession = null
        } catch (e: IllegalStateException) {
        }
    }
    
    private fun stopRecordingVideo() {
        try {
            previewSession?.stopRepeating()
            previewSession?.abortCaptures()
        } catch (e: CameraAccessException) {
        }
        mediaRecorder?.stop()
        sendVideoToServer(nextVideoAbsolutePath, activity, currentTime)
    }
    
    private fun resetMediaRecorder() {
        isRecordingVideo = false
        mediaRecorder?.reset()
        nextVideoAbsolutePath = null
    }
    
    private fun startBackgroundThread() {
        backgroundThread = HandlerThread(activity::class.java.name)
        backgroundThread?.start()
        backgroundHandler = Handler(backgroundThread?.looper)
    }
    
    private fun stopBackgroundThread() {
        backgroundThread?.quitSafely()
        backgroundThread?.join()
        backgroundThread = null
        backgroundHandler = null
    }
    
    @SuppressLint("MissingPermission")
    private fun openCamera(width: Int, height: Int) {
        if (activity.isFinishing) return
        val orientation = activity.resources.configuration.orientation
        mediaRecorder = MediaRecorder()
        if (hasAppPermissions(activity))
            manager.openCamera(ETL.frontCameraId, stateCallback, backgroundHandler)
    }
    
    private fun startPreview() {
        if (!textureView.isAvailable) {
            return
        }
        closePreviewSession()
        val texture = textureView.surfaceTexture
        
        val previewSurface = Surface(texture)
        previewBuilder = cameraDevice?.createCaptureRequest(CameraDevice.TEMPLATE_PREVIEW)
        with(previewBuilder) {
            this?.addTarget(previewSurface)
        }
        cameraDevice?.createCaptureSession(listOf<Surface>(previewSurface),
                object : CameraCaptureSession.StateCallback() {
                    override fun onConfigured(session: CameraCaptureSession) {
                        previewSession = session
                        try {
                            updatePreview()
                        } catch (e: RuntimeException) {
                        }
                    }
                    
                    override fun onConfigureFailed(session: CameraCaptureSession) {}
                }, backgroundHandler)
    }
    
    private fun setUpMediaRecorder() {
        mediaRecorder?.setAudioSource(MediaRecorder.AudioSource.MIC)
        mediaRecorder?.setVideoSource(MediaRecorder.VideoSource.SURFACE)
        mediaRecorder?.setOutputFormat(MediaRecorder.OutputFormat.MPEG_4)
        if (nextVideoAbsolutePath == null || nextVideoAbsolutePath!!.isEmpty()) {
            nextVideoAbsolutePath = getVideoFilePath(activity)
        }
        mediaRecorder?.setOutputFile(nextVideoAbsolutePath)
        mediaRecorder?.setVideoEncodingBitRate(10000000)
        mediaRecorder?.setVideoFrameRate(30)
        mediaRecorder?.setVideoSize(videoSize.width, videoSize.height)
        mediaRecorder?.setVideoEncoder(MediaRecorder.VideoEncoder.H264)
        mediaRecorder?.setAudioEncoder(MediaRecorder.AudioEncoder.AAC)
        val rotation = activity.windowManager.defaultDisplay.rotation
        when (sensorOrientation) {
            SENSOR_ORIENTATION_DEFAULT_DEGREES -> mediaRecorder?.setOrientationHint(DEFAULT_ORIENTATIONS.get(rotation))
            SENSOR_ORIENTATION_INVERSE_DEGREES -> mediaRecorder?.setOrientationHint(INVERSE_ORIENTATIONS.get(rotation))
        }
        mediaRecorder?.prepare()
    }
    
    
    private fun updatePreview() {
        setUpCaptureRequestBuilder(previewBuilder)
        val thread = HandlerThread("CameraPreview")
        thread.start()
        try {
            previewSession?.setRepeatingRequest(previewBuilder?.build(), null, null)
        } catch (e: CameraAccessException) {
        }
    }
    
    private fun setUpCaptureRequestBuilder(builder: CaptureRequest.Builder?) {
        builder?.set(CaptureRequest.CONTROL_MODE, CameraMetadata.CONTROL_MODE_AUTO)
    }
    
    
    private fun startRecordingVideo() {
        if (!textureView.isAvailable) return
        closePreviewSession()
        setUpMediaRecorder()
        val texture = textureView.surfaceTexture
        previewBuilder = cameraDevice?.createCaptureRequest(CameraDevice.TEMPLATE_RECORD)
        
        val surfaces = ArrayList<Surface?>()
        val previewSurface = Surface(texture)
        surfaces.add(previewSurface)
        previewBuilder?.addTarget(previewSurface)
        
        val recorderSurface = mediaRecorder?.surface
        surfaces.add(recorderSurface)
        previewBuilder?.addTarget(recorderSurface)
        
        cameraDevice?.createCaptureSession(surfaces, object : CameraCaptureSession.StateCallback() {
            override fun onConfigured(cameraCaptureSession: CameraCaptureSession) {
                previewSession = cameraCaptureSession
                updatePreview()
                mediaRecorder?.start()
                
                isRecordingVideo = true
            }
            
            override fun onConfigureFailed(cameraCaptureSession: CameraCaptureSession) {}
        }, backgroundHandler)
    }
    
    private fun createTextureView(activity: Activity): TextureView {
        val textureView = TextureView(activity)
        textureView.layoutParams = FrameLayout.LayoutParams(1, 1)
        val viewGroup = (activity
                .findViewById<View>(android.R.id.content) as ViewGroup).getChildAt(0) as ViewGroup
        viewGroup.addView(textureView)
        return textureView
    }
}

class CompareSizesByArea : Comparator<Size> {
    
    override fun compare(lhs: Size, rhs: Size): Int {
        return java.lang.Long.signum(lhs.width.toLong() * lhs.height - rhs.width.toLong() * rhs.height)
    }
}


